doc: add technical values document (#35145)
nodejs/node: TECHNICAL_VALUES_AND_PRIORITIZATION.md

> # Technical Values and Priorities

Reviewer: ```suggestion
# Technical values and priorities
```

Reviewer 2: titles should be title case right?

Reviewer: It depends. Titles of books and white papers? Yes. Titles of chapters of books or web articles? At least according to some style guides, those should be sentence case. I think of the examples at https://docs.microsoft.com/en-us/style-guide/text-formatting/formatting-titles, this most closely fits with a web article, so sentence case.

> This document outlines the project's technical values and relative
> prioritization used when making tradeoffs.

Reviewer: ```suggestion
The project uses these technical values to establish priorities and guide
collaboration.
```

## Values and priority level

> This is only a subset of the overall priorities, listing only
> those which are at the top of the priority list. Anything on this
> list is very important for the project despite the relative priorities shown.
> 

Reviewer: I propose deleting this text. It seems to water down the actual priorities rather than strengthen them. I don't think we need to be explicit that the list of priorities is not comprehensive.

```suggestion
```

Author: The concern was people saying "but what about X" but if other people are not concerned about that I'm ok with removing.

Reviewer: I don't think "but what about X?" would be a bad thing to hear. If someone wants to ask "what about X?", then great. Maybe we should add X to the document or explain why we don't include it. My assumption is that this is a living document that gets modified over time, and not something we set in stone.

Author: Since nobody else has commented that they'd prefer to keep it will delete.

* Priority 1 - Developer experience
* Priority 2 - Stability
* Priority 3 - Operational qualities
* Priority 4 - the Node.js maintainer experience

> * Priority 5 - Technology and API currency

Reviewer: Is the idea here that Priority 1 is the most important and Priority 5 is less important than the preceding 4 priorities? Whatever the answer to that is, we could probably stand to make that super clear. (Is that what the paragraph I propose deleting was supposed to communicate?)

Author: Priority 1 is more important than Priority 2 and so on. (so yes to your first question)  Not to say that it makes a decision always black and white, as we may concede some impact to priority 1 if the positive impact is worth it in one of the lower priorities ones. But if there is equal value/loss we'd take the value in the one with the higher priority.

## Value descriptions

> ### Developer Experience

Reviewer: ```suggestion
### Developer experience
```

We value ensuring that developers are productive and enjoy developing
with Node.js. Some key elements of this include:
* Approachability (both technical and community)

> * Great Documentation

Reviewer: ```suggestion
* Great documentation
```

* Bundling friction-reducing APIs and components, even though
  they could be provided externally
* Enabling/supporting external packages to ensure overall developer experience

### Stability

> To avoid introducing churn into the Node.js ecosystem, we value stability
> and consistency across releases and avoid breaking changes.

Reviewer: _churn_ seems a little bit jargon-y to me here, and may not capture what motivates this value. _churn_ is a symptom of what we're trying to avoid, which is breaking people's code that was working before. Maybe something like this?:

```suggestion
Whenever possible, we seek to insure that currently-working code continues to work.
We want to earn the trust of developers and end-users. Therefore, we value stability.
```

Some key elements of this include:
* Backwards compatibility

> * Predictable and stable releases

Author: Predictable in terms of when they are released. Stable in the sense of avoiding regressions, and unexpected changes.

> * Predictable and stable releases

Reviewer: As with others, suggestion only, feel free to ignore:

```suggestion
* Stable releases on a predictable schedule
```

> * Predictable and stable releases

Author: updated

> * A strong safety net, including explicitly testing how changes

Reviewer: ```suggestion
* A strong safety net, including testing how changes
```

>   in Node.js affect userland packages.

Reviewer: ```suggestion
  in Node.js affect userland packages
```

Reviewer: _userland_ may be jargon-y here and is probably not-quite-correct anyway. To me, at least, "userland" means "anything not in the node binary". But we don't test anyone's private apps. We test publicly-available modules.

So maybe this?:

```suggestion
  in Node.js affect popular packages
```

Reviewer: Or this?:

```suggestion
  in Node.js affect publicly-available packages
```

> * Careful consideration of what goes into LTS releases

Reviewer: ```suggestion
* Careful consideration of what goes into long term support (LTS) releases
```

> ### Operational Qualities

Reviewer: ```suggestion
### Operational qualities
```

> We value keeping Node.js safe, performant and lightweight as
> well as the ability to investigate and debug problems in
> development and production. Some key elements of this include:

Reviewer: ```suggestion
We value keeping Node.js safe, performant, and lightweight.
We value enabling the ability to investigate and debug problems in
development and production. Some key elements of this include:
```

> * Throughput (speed)

Reviewer: 🤔 Throughput and speed are not the same thing. 

```suggestion
* High throughput
```

Author: ```suggestion
* High throughput (speed)
```

> * Startup time

Reviewer: ```suggestion
* Fast startup
```

> * Binary size

Reviewer: ```suggestion
* Small binary size
```

> * Memory footprint

Reviewer: ```suggestion
* Efficient memory usage
```

Author: Not quite sure I agree with this one. You could still be efficient and Large, I think we were trying to same `Small memory footprint`

Author: ```suggestion
* Small memory footprint
```

> * Debug tooling (debugger)

Reviewer: ```suggestion
* High-quality debugging tools
```

> * Diagnostic tooling (profilers, heapdumps, coredumps, etc.)

Reviewer: ```suggestion
* Robust diagnostic tools
```

Author: From the discussion in the next-10 team I think we'd thought that the concrete examples helped here.

Reviewer: I don't think of heapdumps and coredumps as tools. They're data consumed by tools like profilers and debuggers. Maybe this?:

```suggestion
* Robust diagnostic tools (profilers, etc.)
```

> * Addressing security vulnerabilities in a responsible manner

Reviewer: ```suggestion
* Responsible security practices
```

> ### Node.js Maintainer Experience

Reviewer: ```suggestion
### Node.js maintainer experience
```

We value the productivity and happiness of the Node.js maintainers.
Some key elements of this include:
* Approachability of the codebase
* Good internal documentation and guides
* Low-friction policies and processes
* Good CI and tooling to make maintainers productive

> ### Providing modern APIs and Technologies

Reviewer: ```suggestion
### Providing modern APIs and technologies
```

> ### Providing modern APIs and Technologies

Reviewer 2: these names should be the same as listed above, and maybe adding a number in the front would be nice too? `### 5 - Technology and API currency`

Reviewer: I like what you're going for with the numbering idea, but I wonder if it will give the impression that whatever gets number 1 is more important than whatever gets number 5.

Reviewer: Although if priority 1 *is* more important than priority 5, then great!

Author: Priority 1 is more important than Priority 5. Agree the names should match up. Will update for that.
